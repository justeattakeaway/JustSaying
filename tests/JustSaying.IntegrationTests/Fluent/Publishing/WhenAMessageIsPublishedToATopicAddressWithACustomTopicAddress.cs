using JustSaying.Messaging.MessageHandling;
using JustSaying.Models;
using JustSaying.TestingFramework;
using Microsoft.Extensions.DependencyInjection;
using Newtonsoft.Json;

namespace JustSaying.IntegrationTests.Fluent.Publishing;

public class WhenAMessageIsPublishedToATopicAddressWithACustomTopicAddress(ITestOutputHelper outputHelper) : IntegrationTestBase(outputHelper)
{
    [AwsFact]
    public async Task Then_The_Message_Is_Handled()
    {
        // Arrange
        var cancellationToken = CancellationToken.None;
        var topicArn = await GivenAnExistingTopic("uk-simple-message", cancellationToken);
        _ = await GivenAnExistingTopic("us-simple-message", cancellationToken);
        var topicArnTemplate = topicArn.Replace("uk", "{Tenant}");
        var handler = new InspectableHandler<SimpleMessage>();

        var services = GivenJustSaying()
            .ConfigureJustSaying((builder) =>
                builder
                    .Publications((options) =>
                        options.WithTopicArn<SimpleMessage>(topicArnTemplate, (opt) => opt.WithTopicAddress(TenantTopicAddressCustomizer)))
                    .Subscriptions((options) =>
                    {
                        options.ForTopic<SimpleMessage>("uk-simple-message",
                            subscriptionBuilder => { subscriptionBuilder.WithQueueName(UniqueName); });
                        options.ForTopic<SimpleMessage>("us-simple-message",
                            subscriptionBuilder => { subscriptionBuilder.WithQueueName(UniqueName); });
                    }))
            .AddSingleton<IHandlerAsync<SimpleMessage>>(handler);

        string content = Guid.NewGuid().ToString();

        var ukMessage = new SimpleMessage()
        {
            Content = content,
            Tenant = "uk",
        };

        var usMessage = new SimpleMessage()
        {
            Content = content,
            Tenant = "us",
        };

        string json = "";

        await WhenAsync(
            services,
            async (publisher, listener, cancellationToken) =>
            {
                await listener.StartAsync(cancellationToken);
                await publisher.StartAsync(cancellationToken);

                await publisher.PublishAsync(ukMessage, cancellationToken);
                await publisher.PublishAsync(usMessage, cancellationToken);

                // Interrogation has to come after publishing, as per-arn publishers are created on demand
                var listenerJson = JsonConvert.SerializeObject(listener.Interrogate(), Formatting.Indented);
                var publisherJson = JsonConvert.SerializeObject(publisher.Interrogate(), Formatting.Indented);

                json = string.Join($"{Environment.NewLine}{Environment.NewLine}",
                        listenerJson,
                        publisherJson)
                    .Replace(UniqueName, "integrationTestQueueName", StringComparison.Ordinal);

                await Patiently.AssertThatAsync(OutputHelper,
                    () =>
                    {
                        handler.ReceivedMessages.Any(x => x.Content == content && x.Tenant == "uk").ShouldBeTrue();
                        handler.ReceivedMessages.Any(x => x.Content == content && x.Tenant == "us").ShouldBeTrue();
                    });
            });

        json.ShouldMatchApproved(opt => opt.SubFolder("Approvals"));
    }

    private static string TenantTopicAddressCustomizer(string topicArnTemplate, Message message) =>
        topicArnTemplate.Replace("{Tenant}", message.Tenant);
}
