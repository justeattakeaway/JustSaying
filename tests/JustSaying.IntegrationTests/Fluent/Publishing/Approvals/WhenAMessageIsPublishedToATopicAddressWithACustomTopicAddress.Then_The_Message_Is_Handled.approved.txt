{
  "Region": "eu-west-1",
  "Middleware": {
    "Middlewares": [
      {
        "MessageType": "SimpleMessage",
        "QueueName": "integrationTestQueueName",
        "MiddlewareChain": [
          "MessageContextAccessorMiddleware",
          "LoggingMiddleware",
          "StopwatchMiddleware",
          "SqsPostProcessorMiddleware",
          "ErrorHandlerMiddleware",
          "HandlerInvocationMiddleware`1[JustSaying.TestingFramework.SimpleMessage]"
        ]
      }
    ]
  },
  "PublishedMessageTypes": {},
  "SubscriptionGroups": {
    "Groups": [
      {
        "Name": "integrationTestQueueName",
        "ConcurrencyLimit": 10,
        "Multiplexer": {
          "ChannelCapacity": 100,
          "ReaderCount": 2
        },
        "ReceiveBuffers": [
          {
            "BufferSize": 10,
            "QueueName": "integrationTestQueueName",
            "Region": "eu-west-1",
            "Prefetch": 10
          },
          {
            "BufferSize": 10,
            "QueueName": "integrationTestQueueName",
            "Region": "eu-west-1",
            "Prefetch": 10
          }
        ]
      }
    ]
  }
}

{
  "Region": "eu-west-1",
  "Middleware": {
    "Middlewares": []
  },
  "PublishedMessageTypes": {
    "SimpleMessage": {
      "Publishers": {
        "arn:aws:sns:us-east-1:000000000000:uk-simple-message": {
          "Arn": "arn:aws:sns:us-east-1:000000000000:uk-simple-message"
        },
        "arn:aws:sns:us-east-1:000000000000:us-simple-message": {
          "Arn": "arn:aws:sns:us-east-1:000000000000:us-simple-message"
        }
      }
    }
  },
  "SubscriptionGroups": {
    "Groups": []
  }
}